lib/elixir/lib/task/supervised.ex

```diff
@@ -149,99 +149,101 @@ defmodule Task.Supervised do
     parent = self()
 
     # Start a process responsible for translating down messages.
-    {monitor_pid, monitor_ref} = spawn_monitor(fn -> stream_monitor(parent) end)
+    {:trap_exit, trap_exit} = Process.info(self(), :trap_exit)
+    monitor_pid = spawn_link(fn -> stream_monitor(parent, mfa, spawn, trap_exit) end)
+    monitor_ref = Process.monitor(monitor_pid)
     send(monitor_pid, {parent, monitor_ref})
 
     stream_reduce(acc, max_concurrency, 0, 0, %{}, next,
-                  reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+                  reducer, monitor_pid, monitor_ref, timeout)
   end
 
-  defp stream_reduce({:halt, acc}, _max, _spawned, _delivered, waiting, next,
-                     _reducer, _mfa, _spawn, monitor_pid, monitor_ref, timeout) do
+  defp stream_reduce({:halt, acc}, _max, _spawned, _delivered, _waiting, next,
+                     _reducer, monitor_pid, monitor_ref, timeout) do
     is_function(next) && next.({:halt, []})
-    stream_close(waiting, monitor_pid, monitor_ref, timeout)
+    stream_close(monitor_pid, monitor_ref, timeout)
```

Comment:
Note quite related to this PR but should we call `stream_close` before `next`? It seems like `next` could raise on cleanup? We should test this.